Machine Learning I exercise, done by Cedric Prieels.

In [80]:
import keras
keras.__version__

'2.2.4'

# Introduction to ConvNets: Classifying handwritten numbers


Let's take a look at a simple example of a convnet. We will use it to classify the MNIST dataset, which is an open dataset containing handwritten numbers. 

![Handwritten numbers from the MNIST dataset](http://corochann.com/wp-content/uploads/2017/02/mnist_plot.png)

Let's create a first basic convnet. It's a stack of 'Conv2D' and 'MaxPooling2D' layers. 
The important thing to note is that a convnet takes as input tensors of size `(image_height, image_width, image_channels)`. 
To do this we must first find out the size of the images in our dataset. 

The network must have the following layers:

- A convolutional layer (Conv2D) with 32 3x3 filters and relu activation. In this first layer you must indicate the size of the input (input_shape).
- A second layer of Max Pooling (MaxPooling2D) of 2x2
- A third convolutional layer with 64 3x3 filters and light activation
- A fourth layer of 2x2 Max Pooling (MaxPooling2D)
- A fifth convolutional layer of 64 3x3 filters and light activation

You'll know you've done it right when the model.summary() output is:

![imagen_output.png](https://github.com/laramaktub/MachineLearningI/blob/master/imagen_output.png?raw=true)


In [60]:
from keras import layers
from keras import models

model = models.Sequential()
model.add(layers.Conv2D(32, (3, 3), activation='relu', input_shape=(28, 28, 1))) #The minst dataset images have a size of 28x28px
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='relu')) #This time, we use 64 filters of size 3x3
model.add(layers.MaxPooling2D((2, 2)))
model.add(layers.Conv2D(64, (3, 3), activation='linear')) #Last layer

model.summary() #We have a total of 55744 parameters in this convolutional network!

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_10 (Conv2D)           (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_7 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_11 (Conv2D)           (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_8 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_12 (Conv2D)           (None, 3, 3, 64)          36928     
Total params: 55,744
Trainable params: 55,744
Non-trainable params: 0
_________________________________________________________________


You can see above that the output of each Conv2D and MaxPooling2D layer is a 3D tensor of dimensions (height, width, channels). The width and height tend to decrease as we go deeper into the network. The number of channels is controlled by the first argument passed to the Conv2D layers (e.g. 32 or 64).

The next step would be to give our last tensor (of dimensions (3, 3, 64)) as input to a densely connected network. These classifiers process vectors, which are 1D, while our output is a 3D tensor. So first we will have to flatten our 3D output and convert it to 1D and then add a few dense layers:

- First flatten the output.
- Add a first layer of 64 neurons and relu activation
- Add a last layer of 10 neurons (as many as you can sort) and software activation
- You'll know you've done well when the summary looks like this:

![imagen_output_flat.png](https://github.com/laramaktub/MachineLearningI/blob/master/imagen_output_flat.png?raw=true)

In [61]:
from keras.layers import Dense, Activation, Flatten
model.add(Flatten())
model.add(Dense(64))
model.add(Activation('relu'))
model.add(Dense(10))
model.add(Activation('softmax'))

model.summary() #Check the results

_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_10 (Conv2D)           (None, 26, 26, 32)        320       
_________________________________________________________________
max_pooling2d_7 (MaxPooling2 (None, 13, 13, 32)        0         
_________________________________________________________________
conv2d_11 (Conv2D)           (None, 11, 11, 64)        18496     
_________________________________________________________________
max_pooling2d_8 (MaxPooling2 (None, 5, 5, 64)          0         
_________________________________________________________________
conv2d_12 (Conv2D)           (None, 3, 3, 64)          36928     
_________________________________________________________________
flatten_4 (Flatten)          (None, 576)               0         
_________________________________________________________________
dense_3 (Dense)              (None, 64)                36928     
__________

As you can see, our dimensional output `(3, 3, 64)` has been flattened into a vector of dimension `(576,)`, before entering the two dense layers.

We are now going to train our network with the images from the MNIST dataset.

We then load the dataset and put it into vectors: train_images, train_labels, test_images, test_labels

Before you continue, print:

- What is the size of the training dataset?
- What does the training dataset look like?
- What do the training labels look like?
- Print the fourth image of the training dataset


In [62]:
from keras.datasets import mnist
from keras.utils import to_categorical

(train_images, train_labels), (test_images, test_labels) = mnist.load_data()

In [63]:
print("Size of the training dataset: " + str(len(train_images)))
print(train_images, train_labels) #We have an multidimentional array corresponding to the pixels of each image (28x28), and then a label (a number between 0 and 9) associated to each image
print("\nFourth image pixels: (size = " +str(len(train_images[4])) + "x" + str(len(train_images[4][0])) + "px)")
print(train_images[3])

Size of the training dataset: 60000
[[[0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]
  ...
  [0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]]

 [[0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]
  ...
  [0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]]

 [[0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]
  ...
  [0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]]

 ...

 [[0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]
  ...
  [0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]]

 [[0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]
  ...
  [0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]]

 [[0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]
  ...
  [0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]
  [0 0 0 ... 0 0 0]]] [5 0 4 ... 5 6 8]

Fourth image pixels: (size = 28x28px)
[[  0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0   0
    0   0   0   0   0   0   0   0   0   0]
 [  0   0   0   

Next you will give the appropriate shape to the training and test datasets in order to put them into the neural network. Convert the labels, which right now are numbers, into their categorical form.

In [64]:
train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255 #Rescale the images

from keras.utils import to_categorical
#Change the name of the variable to avoid overwriting the previous one in case we need it
num_classes = 10
train_labels_cat = to_categorical(train_labels, num_classes)
print(train_labels_cat)
test_labels_cat = to_categorical(test_labels, num_classes)

[[0. 0. 0. ... 0. 0. 0.]
 [1. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 ...
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 0. 0.]
 [0. 0. 0. ... 0. 1. 0.]]


Compile the model indicating what the training data and its labels are. Using the optimizer "rmsprop" and as a loss function use the categorical cross entropy.
Then train the model for 5 epochs and a batch size of 64.

In [65]:
model.compile(optimizer="rmsprop", loss="categorical_crossentropy", metrics=['accuracy'])
model.fit(x=train_images, y=train_labels_cat, epochs=5, batch_size=64)

Epoch 1/5
60000/60000 [==============================] - 54s 907us/step - loss: 0.1535 - acc: 0.9513
Epoch 2/5
60000/60000 [==============================] - 57s 955us/step - loss: 0.0496 - acc: 0.9843
Epoch 3/5
60000/60000 [==============================] - 56s 930us/step - loss: 0.0356 - acc: 0.9891
Epoch 4/5
60000/60000 [==============================] - 58s 971us/step - loss: 0.0280 - acc: 0.9910
Epoch 5/5
60000/60000 [==============================] - 55s 917us/step - loss: 0.0227 - acc: 0.9930


Let's evaluate the model with the test images:

In [66]:
test_loss, test_acc = model.evaluate(test_images, test_labels_cat)
print(test_loss, test_acc) #Accuracy of 98.97% achieved with the test dataset!

10000/10000 [==============================] - 5s 483us/step
0.9897


Create an image with a handwritting number and check the prediction. Try with several numbers ...does it work properly?

In [79]:
from keras.preprocessing import image
import numpy as np
from matplotlib import pyplot as plt

img_width=28
img_height=28

def myPredict(filename, name):
    img = image.load_img(filename, target_size=(img_width, img_height),grayscale=True)
    x = image.img_to_array(img)
    x = np.expand_dims(x, axis=0)
    print(name + ": " + str(model.predict_classes(x)))

myPredict("zero.png", "Zero")    
myPredict("one.png", "One")
myPredict("two.png", "Two")    
myPredict("four.png", "Four")    
myPredict("eight.png", "Eight")    

Zero: [0]
One: [3]
Two: [2]
Four: [4]
Eight: [8]


We can see that four of my five numbers were correctly classified: the 0, the 2, the 4 and the 8. For some reason, it appears the 1 I wrote has been confused with a 3. Using this very limited dataset, I get an accuracy of 80%.

In [77]:
model.save('net_numbers.h5')

Load the model that you just saved and make a prediction (predict_classes) with the number you just generated. 
